File: src/data/courses/dsa/stacks/stack-basic-applications/evaluate-nested-expression/code/cpp/EvaluateNestedExpression.cpp
// ============================================================
// Evaluate Nested Expression
// ============================================================

#include <string>
#include <stack>
#include <vector>
#include <cctype>
using namespace std;

// ============================================================
// Solution: Stack-Based Evaluation
// Time: O(N) | Space: O(N)
// ============================================================
class Solution {
public:
    int evaluate(string s) {
        // Stack of (operator, operands)
        stack<pair<string, vector<int>>> st;

        for (int i = 0; i < s.size(); i++) {
            
            // Parse operator (letters)
            if (isalpha(s[i])) {
                string op;
                while (i < s.size() && isalpha(s[i])) {
                    op += s[i++];
                }

                // Validate operator
                if (op != "add" && op != "sub" && op != "mult") {
                    return -1;  // Invalid operator
                }

                st.push({op, {}});
                i--;  // Adjust for loop increment
            }

            // Parse number (digits)
            else if (isdigit(s[i])) {
                // Number without operator frame → invalid
                if (st.empty()) return -1;

                int num = 0;
                while (i < s.size() && isdigit(s[i])) {
                    num = num * 10 + (s[i++] - '0');
                }

                // Add operand to current frame
                st.top().second.push_back(num);
                i--;  // Adjust for loop increment
            }

            // Closing bracket → compute result
            else if (s[i] == ')') {
                // No frame to close → invalid
                if (st.empty()) return -1;

                auto [op, nums] = st.top();
                st.pop();

                // Validate: exactly 2 operands
                if (nums.size() != 2) return -1;

                // Compute result based on operator
                int result;
                if (op == "add") {
                    result = nums[0] + nums[1];
                } else if (op == "sub") {
                    result = nums[0] - nums[1];
                } else {  // mult
                    result = nums[0] * nums[1];
                }

                // Pass result to parent or return if outermost
                if (!st.empty()) {
                    st.top().second.push_back(result);
                } else {
                    // Check if we've consumed entire string
                    if (i == s.size() - 1) {
                        return result;
                    } else {
                        return -1;  // Extra characters after result
                    }
                }
            }

            // Ignore commas, spaces, opening parentheses
            else if (s[i] == ',' || s[i] == ' ' || s[i] == '(') {
                continue;
            }

            // Invalid character
            else {
                return -1;
            }
        }

        // If stack not empty → unfinished expression
        return -1;
    }
};

File: src/data/courses/dsa/stacks/stack-basic-applications/evaluate-nested-expression/code/python/evaluate_nested_expression.py
# ============================================================
# Evaluate Nested Expression
# ============================================================

# ============================================================
# Solution: Stack-Based Evaluation
# Time: O(N) | Space: O(N)
# ============================================================
class Solution:
    def evaluate(self, s: str) -> int:
        # Stack of [operator, operands]
        stack = []

        i = 0
        while i < len(s):
            c = s[i]
            
            # Parse operator (letters)
            if c.isalpha():
                op = ""
                while i < len(s) and s[i].isalpha():
                    op += s[i]
                    i += 1

                # Validate operator
                if op not in ["add", "sub", "mult"]:
                    return -1  # Invalid operator

                stack.append([op, []])
                continue  # i already incremented

            # Parse number (digits)
            elif c.isdigit():
                # Number without operator frame → invalid
                if not stack:
                    return -1

                num = 0
                while i < len(s) and s[i].isdigit():
                    num = num * 10 + int(s[i])
                    i += 1

                # Add operand to current frame
                stack[-1][1].append(num)
                continue  # i already incremented

            # Closing bracket → compute result
            elif c == ')':
                # No frame to close → invalid
                if not stack:
                    return -1

                op, nums = stack.pop()

                # Validate: exactly 2 operands
                if len(nums) != 2:
                    return -1

                # Compute result based on operator
                if op == "add":
                    result = nums[0] + nums[1]
                elif op == "sub":
                    result = nums[0] - nums[1]
                else:  # mult
                    result = nums[0] * nums[1]

                # Pass result to parent or return if outermost
                if stack:
                    stack[-1][1].append(result)
                else:
                    # Check if we've consumed entire string
                    if i == len(s) - 1:
                        return result
                    else:
                        return -1  # Extra characters after result

            # Ignore commas, spaces, opening parentheses
            elif c in [',', ' ', '(']:
                pass

            # Invalid character
            else:
                return -1
            
            i += 1

        # If stack not empty → unfinished expression
        return -1
